File: Robotics_Kinematics_and_Dynamics_Course/Chapter6/Lesson5/Chapter6_Lesson5.java
import org.ejml.simple.SimpleMatrix;

public class RedundantIK {

    public static SimpleMatrix dampedPseudoinverse(SimpleMatrix J, double lambda) {
        int m = J.numRows();
        SimpleMatrix JJt = J.mult(J.transpose());
        SimpleMatrix I = SimpleMatrix.identity(m);
        SimpleMatrix inv = JJt.plus(I.scale(lambda * lambda)).invert();
        return J.transpose().mult(inv);
    }

    public static SimpleMatrix planar3Jacobian(double[] q, double[] L) {
        double q1 = q[0], q2 = q[1], q3 = q[2];
        double l1 = L[0], l2 = L[1], l3 = L[2];

        double s1 = Math.sin(q1), c1 = Math.cos(q1);
        double s12 = Math.sin(q1 + q2), c12 = Math.cos(q1 + q2);
        double s123 = Math.sin(q1 + q2 + q3), c123 = Math.cos(q1 + q2 + q3);

        SimpleMatrix J = new SimpleMatrix(2, 3);
        J.set(0, 0, -l1 * s1 - l2 * s12 - l3 * s123);
        J.set(1, 0,  l1 * c1 + l2 * c12 + l3 * c123);
        J.set(0, 1, -l2 * s12 - l3 * s123);
        J.set(1, 1,  l2 * c12 + l3 * c123);
        J.set(0, 2, -l3 * s123);
        J.set(1, 2,  l3 * c123);
        return J;
    }

    public static SimpleMatrix jointLimitGradient(double[] q,
                                                  double[] qMin,
                                                  double[] qMax) {
        int n = q.length;
        SimpleMatrix grad = new SimpleMatrix(n, 1);
        for (int i = 0; i < n; ++i) {
            double qMid = 0.5 * (qMin[i] + qMax[i]);
            double span = qMax[i] - qMin[i];
            double normed = (q[i] - qMid) / span;
            grad.set(i, 0, normed / span);
        }
        return grad;
    }

    public static SimpleMatrix redundancyResolutionStep(double[] q,
                                                        double[] xdotDes,
                                                        double[] L,
                                                        double[] qMin,
                                                        double[] qMax,
                                                        double alpha,
                                                        double lambda) {
        SimpleMatrix J = planar3Jacobian(q, L);
        SimpleMatrix Jpinv = dampedPseudoinverse(J, lambda);

        SimpleMatrix xdot = new SimpleMatrix(2, 1, true, xdotDes);
        SimpleMatrix qdot0 = Jpinv.mult(xdot);

        SimpleMatrix I = SimpleMatrix.identity(J.numCols());
        SimpleMatrix N = I.minus(Jpinv.mult(J));

        SimpleMatrix gradH = jointLimitGradient(q, qMin, qMax);
        SimpleMatrix qdotH = N.mult(gradH).scale(-alpha);

        return qdot0.plus(qdotH);
    }

    public static void main(String[] args) {
        double[] L = {0.4, 0.3, 0.2};
        double[] q = {0.0, 0.3, -0.2};
        double[] qMin = {-Math.PI, -Math.PI, -Math.PI};
        double[] qMax = { Math.PI,  Math.PI,  Math.PI};
        double[] xdotDes = {0.05, 0.0};

        SimpleMatrix qdot = redundancyResolutionStep(q, xdotDes,
                                                     L, qMin, qMax,
                                                     0.2, 1e-3);

        System.out.println("qdot = ");
        qdot.print();
    }
}
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter6/Lesson5/Chapter6_Lesson5.py
import numpy as np

def planar3_fk(q, link_lengths):
    """
    Simple planar 3R forward kinematics returning end-effector position.
    q: array-like, shape (3,)
    link_lengths: array-like, shape (3,)
    """
    q1, q2, q3 = q
    l1, l2, l3 = link_lengths
    c1 = np.cos(q1); s1 = np.sin(q1)
    c12 = np.cos(q1 + q2); s12 = np.sin(q1 + q2)
    c123 = np.cos(q1 + q2 + q3); s123 = np.sin(q1 + q2 + q3)

    x = l1 * c1 + l2 * c12 + l3 * c123
    y = l1 * s1 + l2 * s12 + l3 * s123
    return np.array([x, y])

def planar3_jacobian(q, link_lengths):
    """
    Jacobian J(q) mapping joint velocities to end-effector linear velocity.
    Shape: (2, 3).
    """
    q1, q2, q3 = q
    l1, l2, l3 = link_lengths
    s1 = np.sin(q1); c1 = np.cos(q1)
    s12 = np.sin(q1 + q2); c12 = np.cos(q1 + q2)
    s123 = np.sin(q1 + q2 + q3); c123 = np.cos(q1 + q2 + q3)

    # Partial derivatives of x, y with respect to q1, q2, q3
    # x = l1 c1 + l2 c12 + l3 c123
    # y = l1 s1 + l2 s12 + l3 s123
    J = np.zeros((2, 3))
    J[0, 0] = -l1 * s1 - l2 * s12 - l3 * s123
    J[1, 0] =  l1 * c1 + l2 * c12 + l3 * c123
    J[0, 1] = -l2 * s12 - l3 * s123
    J[1, 1] =  l2 * c12 + l3 * c123
    J[0, 2] = -l3 * s123
    J[1, 2] =  l3 * c123
    return J

def damped_pseudoinverse(J, lam=1e-3):
    """
    Damped pseudoinverse: J^T (J J^T + lam^2 I)^(-1).
    Suitable for m <= n and full (or nearly full) row rank.
    """
    m, n = J.shape
    JJt = J @ J.T
    return J.T @ np.linalg.inv(JJt + (lam ** 2) * np.eye(m))

def joint_limit_cost(q, q_min, q_max, w=None):
    """
    Quadratic joint-limit cost H(q).
    """
    q_mid = 0.5 * (q_min + q_max)
    span = q_max - q_min
    if w is None:
        w = np.ones_like(q)
    normed = (q - q_mid) / span
    return 0.5 * np.sum(w * normed * normed)

def joint_limit_grad(q, q_min, q_max, w=None):
    """
    Gradient of the joint-limit cost with respect to q.
    """
    q_mid = 0.5 * (q_min + q_max)
    span = q_max - q_min
    if w is None:
        w = np.ones_like(q)
    normed = (q - q_mid) / span
    return w * normed / (span)

def redundancy_resolution_step(q, xdot_des, link_lengths,
                               q_min, q_max,
                               alpha=0.1, lam=1e-3):
    """
    One Euler integration step of redundancy-resolved IK.

    q: current joint configuration, shape (3,)
    xdot_des: desired end-effector velocity (2,)
    """
    J = planar3_jacobian(q, link_lengths)
    J_pinv = damped_pseudoinverse(J, lam=lam)

    # Primary task
    qdot_0 = J_pinv @ xdot_des

    # Null-space projector
    N = np.eye(J.shape[1]) - J_pinv @ J

    # Secondary objective: joint-limit avoidance
    gradH = joint_limit_grad(q, q_min, q_max)
    qdot_H = -alpha * (N @ gradH)

    qdot = qdot_0 + qdot_H
    return qdot

# Example usage (single time step)
if __name__ == "__main__":
    link_lengths = np.array([0.4, 0.3, 0.2])
    q = np.array([0.0, 0.3, -0.2])
    q_min = np.array([-np.pi, -np.pi, -np.pi])
    q_max = np.array([ np.pi,  np.pi,  np.pi])

    # Track a small velocity in x direction
    xdot_des = np.array([0.05, 0.0])

    qdot = redundancy_resolution_step(q, xdot_des, link_lengths,
                                      q_min, q_max,
                                      alpha=0.2, lam=1e-3)
    q_next = q + 0.01 * qdot  # simple Euler step
    print("qdot:", qdot)
    print("q_next:", q_next)
      
